cpp/src/arrow/memory_pool.cc

```diff
@@ -201,4 +201,38 @@ int64_t LoggingMemoryPool::max_memory() const {
   std::cout << "max_memory: " << mem << std::endl;
   return mem;
 }
+
+ProxyMemoryPool::ProxyMemoryPool(MemoryPool* pool) : pool_(pool) {}
+
+Status ProxyMemoryPool::Allocate(int64_t size, uint8_t** out) {
+  Status s = pool_->Allocate(size, out);
+  proxy_bytes_allocated_ += size;
+  return s;
+}
+
+Status ProxyMemoryPool::Reallocate(int64_t old_size, int64_t new_size, uint8_t** ptr) {
+  Status s = pool_->Reallocate(old_size, new_size, ptr);
+  proxy_bytes_allocated_ += new_size - old_size;
+  return s;
+}
+
+void ProxyMemoryPool::Free(uint8_t* buffer, int64_t size) {
+  pool_->Free(buffer, size);
+  proxy_bytes_allocated_ -= size;
+}
+
+int64_t ProxyMemoryPool::bytes_allocated() const {
```

Comment:
Just realised that `bytes_allocated` refers to the parent Pool. I would have expected that it would return only the allocation of the currrent scope (i.e. `proxy_bytes_allocated_`).